dist/PreserveModuleNamePlugin.js

```diff
@@ -1,65 +1,88 @@
-    for (let root of roots) {
-        let relative = path.relative(root, resource);
-        if (!relative.startsWith('..'))
-            return relative;
-    }
-    return null;
-}
-function fixNodeModule(module, allModules) {
-    if (!/\bnode_modules\b/i.test(module.resource))
-        return null;
-    // The problem with node_modules is that often the root of the module is not /node_modules/my-lib
-    // Webpack is going to look for `main` in `project.json` to find where the main file actually is.
-    // And this can of course be configured differently with `resolve.alias`, `resolve.mainFields` & co.
-    // Our best hope is that the file was not required as a relative path, then we can just preserve that.
-    if (!module.rawRequest.startsWith("."))
-        return module.rawRequest;
-    // Otherwise we need to build the relative path from the module root, which as explained above is hard to find.
-    // Ideally we could use webpack resolvers, but they are currently async-only, which can't be used in before-modules-id
-    // See https://github.com/webpack/webpack/issues/1634
-    // Instead, we'll look for the root library module, because it should have been requested somehow and work from there.
-    // Note that the negative lookahead (?!.*node_modules) ensures that we only match the last node_modules/ folder in the path,
-    // in case the package was located in a sub-node_modules (which can occur in special circumstances).
-    let name = /\bnode_modules[\\/](?!.*\bnode_modules\b)([^\\/]*)/i.exec(module.resource)[1];
-    let entry = allModules.find(m => m.rawRequest === name);
-    if (!entry)
-        throw new Error("PreserveModuleNamePlugin: Unable to find root of module " + name);
-    return name + "/" + path.relative(path.dirname(entry.resource), module.resource);
-}
+"use strict";
+const path = require("path");
+exports.preserveModuleName = Symbol();
+// This plugins preserves the module names of IncludeDependency and 
+// AureliaDependency so that they can be dynamically requested by 
+// aurelia-loader.
+// All other dependencies are handled by webpack itself and don't
+// need special treatment.
+class PreserveModuleNamePlugin {
+    apply(compiler) {
+        compiler.plugin("compilation", compilation => {
+            compilation.plugin("before-module-ids", modules => {
+                let { modules: roots, extensions, alias } = compilation.options.resolve;
+                roots = roots.map(x => path.resolve(x));
+                const normalizers = extensions.map(x => new RegExp(x.replace(/\./g, "\\.") + "$", "i"));
+                for (let module of getPreservedModules(modules)) {
+                    let relative = fixNodeModule(module, modules) ||
+                        makeModuleRelative(roots, module.resource) ||
+                        aliasRelative(alias, module.resource);
+                    // TODO: we should probably report an error if we can't figure out a name for an Aurelia dependency.

```

Comment:
Any file that is not in one of your `resolve.module` folders, that is not inside a `resolve.alias` folder or that is a `node_module` that was requested with a relative path but whose root module is not part of the dependencies.

The first two cases can happen if you create a relative path that `..` outside of your app root.

The node modules case could happen if you never request the root but a specific file like `module-A/xyz` and this file requests a relative file like `./stuff`. It's unlikely as `module-A/xyz` would usually not work (except it does thanks to my plugins). Another possibility is a relative path from the app that lead directly to a module file like `../../node_modules/module-A/xyz`. Not a great idea anyway.

Since I wrote those functions I've changed my strategy and I think that I could handle those node_modules cases properly.

Another case I won't handle properly is if the node_modules "main" is in folder dist, but then requests a relative file that is outside this folder, similar to what I said for the app. source code.

Those are all unlikely and bad practices anyway, but they're one reason why I don't try to name modules when it's not necessary.